__tests__/login/oidc/TokenRequester.spec.ts

```diff
@@ -0,0 +1,244 @@
+      }
+    );
+  });
+
+  it("Fails elegantly if the idp returns a bad value", async () => {
+    setUpMockedReturnValues({
+      responseBody: JSON.stringify({
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        id_token: "ohNoThereIsNoAccessToken"
+      })
+    });
+    const TokenRefresher = getTokenRequester();
+    await expect(
+      /* eslint-disable @typescript-eslint/camelcase */
+      TokenRefresher.request("global", {
+        grant_type: "refresh_token",
+        refresh_token: "thisIsARefreshToken"
+      })
+      /* eslint-enable @typescript-eslint/camelcase */
+    ).rejects.toThrowError("IDP token route returned an invalid response.");
+  });
+
+  it("Fails elegantly if the issuer does not support refresh tokens", async () => {
+    setUpMockedReturnValues({
+      issuerConfig: {
+        ...IssuerConfigFetcherFetchConfigResponse,
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        grantTypesSupported: ["id_token"]
+      }
+    });
+    const TokenRefresher = getTokenRequester();
+    await expect(
+      /* eslint-disable @typescript-eslint/camelcase */
+      TokenRefresher.request("global", {
+        grant_type: "refresh_token",
+        refresh_token: "thisIsARefreshToken"
+      })
+      /* eslint-enable @typescript-eslint/camelcase */
+    ).rejects.toThrowError(
+      "The issuer https://idp.com does not support the refresh_token grant"
+    );
+  });
+
+  it("Fails elegantly if the issuer does not have a token endpoint", async () => {
+    setUpMockedReturnValues({
+      issuerConfig: {
+        ...IssuerConfigFetcherFetchConfigResponse,
+        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+        // @ts-ignore This is ignored to test an edge case
+        tokenEndpoint: null,
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        grantTypesSupported: ["refresh_token"]
+      }
+    });
+    const TokenRefresher = getTokenRequester();
+    await expect(
+      /* eslint-disable @typescript-eslint/camelcase */
+      TokenRefresher.request("global", {
+        grant_type: "refresh_token",
+        refresh_token: "thisIsARefreshToken"
+      })
+      /* eslint-enable @typescript-eslint/camelcase */
+    ).rejects.toThrowError(
+      "This issuer https://idp.com does not have a token endpoint"
+    );
+  });
+
+  it("Fails elegantly if the access token does not have a sub claim", async () => {
+    setUpMockedReturnValues({
+      jwt: {
+        iss: "https://idp.com"
+      }
+    });
+    const TokenRefresher = getTokenRequester();
+    await expect(
+      /* eslint-disable @typescript-eslint/camelcase */
+      TokenRefresher.request("global", {
+        grant_type: "refresh_token",
+        refresh_token: "thisIsARefreshToken"
+      })
+      /* eslint-enable @typescript-eslint/camelcase */
+    ).rejects.toThrowError("The idp returned a bad token without a sub.");
```

Comment:
Error message here - why is it 'idp' and not 'IDP' (and in fact I prefer 'IdP', but more important to just be consistent everywhere!). I'd reword slightly too to: "The IdP returned a bad token: missing 'sub' claim."